File: flutter/lib/api/internal/event/digia_analytics_sink.dart
import '../../../src/analytics/analytics_service.dart';
import '../../models/cep_trigger_payload.dart';
import '../campaign/campaign_store.dart';
import 'engage_analytics_event.dart';

/// Delivers rich [EngageAnalyticsEvent]s to Digia's first-party analytics
/// backend.
///
/// Resolves the campaign from [CampaignStore] by `campaignKey` for attribution
/// context (campaign id/type live on the campaign, not the trigger payload),
/// then hands the event to [DigiaAnalyticsService.capture], which merges
/// [EngageAnalyticsEvent.properties] with the static context.
class DigiaAnalyticsSink {
  final DigiaAnalyticsService _analytics;
  final CampaignStore _store;

  DigiaAnalyticsSink(this._analytics, this._store);

  void deliver(EngageAnalyticsEvent event, CEPTriggerPayload payload) {
    final campaign = _store.find(payload.campaignKey);
    _analytics.capture(
      event,
      payload,
      campaignType: campaign?.campaignType,
      campaignId: campaign?.id,
    );
  }
}

File: android/digia-engage/src/main/java/com/digia/engage/internal/event/DigiaAnalyticsSink.kt
package com.digia.engage.internal.event

import com.digia.engage.CEPTriggerPayload
import com.digia.engage.internal.analytics.AnalyticsService
import com.digia.engage.internal.model.CampaignModel

/**
 * Delivers rich [EngageAnalyticsEvent]s to Digia's first-party analytics backend.
 *
 * Resolves the campaign from the store by `campaignKey` for attribution context
 * (campaign id/type live on the [CampaignModel], not the trigger payload), then
 * hands the event to [AnalyticsService], which hoists [EngageAnalyticsEvent.columns]
 * to top level and nests [EngageAnalyticsEvent.properties].
 */
internal class DigiaAnalyticsSink(
    private val getAnalyticsService: () -> AnalyticsService?,
    private val getCampaign: (String) -> CampaignModel?,
) {
    fun deliver(event: EngageAnalyticsEvent, payload: CEPTriggerPayload) {
        val svc = getAnalyticsService() ?: return
        val campaign = getCampaign(payload.campaignKey)
        svc.capture(event, payload, campaignId = campaign?.id, campaignType = campaign?.campaignType)
    }
}
